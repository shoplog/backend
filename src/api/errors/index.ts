export * from './http.error';
export * from './unauthorized.error';
