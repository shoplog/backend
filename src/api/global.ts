import { VPICController } from 'src/api/controllers/vpic.controller';
import { VPICDatabase } from 'src/common/initializers/database';
import {
	LookupRepository,
	MakeRepository,
	ModelRepository,
	VinRepository,
	YearRepository,
} from 'src/data/vpic/repositories';
import { VPICService } from 'src/domains/vpic/services/vpic.service';

const vinRepository = new VinRepository(VPICDatabase);
const yearRepository = new YearRepository(VPICDatabase);
const makeRepository = new MakeRepository(VPICDatabase);
const modelRepository = new ModelRepository(VPICDatabase);
const lookupRepository = new LookupRepository(VPICDatabase);
const vpicService = new VPICService(vinRepository, yearRepository, makeRepository, modelRepository, lookupRepository);
const vpicController = new VPICController(vpicService);

export { vpicController };
