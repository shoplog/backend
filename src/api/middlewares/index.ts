export * from './error-handler.middleware';
export * from './contex-wrap.middleware';
export * from './check-jwt.middleware';
