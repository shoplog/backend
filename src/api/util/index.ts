export * from 'src/api/util/helpers';
