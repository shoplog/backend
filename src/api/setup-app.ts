import { apiReference } from '@scalar/express-api-reference';
import compression from 'compression';
import express, { Request, json } from 'express';
import { middleware as OpenApiValidatorMiddleware } from 'express-openapi-validator';
import { NotFound } from 'express-openapi-validator/dist/openapi.validator';
import helmet from 'helmet';
import pinoHttp from 'pino-http';
import { CUSTOM_HEADERS } from 'src/api/constants/headers';
import { errorHandlerMiddleware } from 'src/api/middlewares';
import { createVehiclesRoutes } from 'src/api/routes/vehicles.route';
import { logger } from 'src/common/initializers/logger';

const apiSpec = 'data/openapi/v1.yml';

export const setupApp = async () => {
	const app = express();

	app.use(
		helmet({
			// Disable CSP because it's not needed for this API only project.
			contentSecurityPolicy: false,
		})
	);

	// For when the app is behind a proxy, like in a docker container or in a kubernetes cluster.
	app.set('trust proxy', 1);

	app.use(compression());

	app.use(
		'/api-docs',
		apiReference({
			spec: {
				content: apiSpec,
			},
		})
	);

	// Generate request id that will correlate all logs for a single request.
	app.use((req, res, next) => {
		req.id = crypto.randomUUID();
		res.set(CUSTOM_HEADERS.RequestId, req.id);
		next();
	});

	app.use(
		pinoHttp({
			logger,
			quietReqLogger: true,
			genReqId: (req) => req.id,
		})
	);

	app.use(json());

	app['get']('/', (req, res) => {
		res.status(200).end();
	});

	app.use('/api/v1', await createVehiclesRoutes());

	app.use(
		OpenApiValidatorMiddleware({
			apiSpec,
			validateApiSpec: false,
			validateRequests: true,
			validateResponses: true,
		})
	);

	app.use((req: Request) => {
		throw new NotFound({ path: req.url, message: 'Not Found' });
	});

	app.use(errorHandlerMiddleware());

	return app;
};
