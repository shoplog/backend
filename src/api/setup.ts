import { apiReference } from '@scalar/express-api-reference';
import compression from 'compression';
import cors from 'cors';
import express, { json } from 'express';
import { middleware as OpenApiValidatorMiddleware } from 'express-openapi-validator';
import { OpenApiSpecLoader } from 'express-openapi-validator/dist/framework/openapi.spec.loader';
import { NotFound, OpenAPIV3 } from 'express-openapi-validator/dist/framework/types';
import 'express-openapi-validator/dist/middlewares/parsers/schema.parse';
import helmet from 'helmet';
import pinoHttp from 'pino-http';
import { OPEN_API_SPEC_FILE_PATH } from 'src/api/constants/files';
import { CUSTOM_HEADERS } from 'src/api/constants/headers';
import { Dependencies } from 'src/api/global';
import { errorHandlerMiddleware } from 'src/api/middlewares';
import { createVPICRoutes } from 'src/api/routes/vpic.routes';
import { logger } from 'src/common/initializers/logger';

export const loadOpenApiSpec = async (): Promise<OpenAPIV3.Document> => {
	const loader = new OpenApiSpecLoader({
		apiDoc: OPEN_API_SPEC_FILE_PATH,
		validateApiSpec: true,
	});

	const { apiDoc: apiSpec } = await loader.load();

	return apiSpec;
};

export const createApp = async (dependencies: Dependencies) => {
	const app = express();
	const apiSpec = await loadOpenApiSpec();

	app.use(
		helmet({
			// Disable CSP because it's not needed for this API only project.
			contentSecurityPolicy: false,
		})
	);

	// For when the app is behind a proxy, like in a docker container or in a kubernetes cluster.
	app.set('trust proxy', 1);

	app.use(compression());

	app.use(cors());

	app.get('/docs/openapi.json', (_req, res) => {
		res.json(apiSpec);
	});

	app.use(
		'/docs',
		apiReference({
			spec: {
				content: apiSpec,
			},
		})
	);

	// Generate request id that will correlate all logs for a single request.
	app.use((req, res, next) => {
		req.id = crypto.randomUUID();
		res.set(CUSTOM_HEADERS.RequestId, req.id);
		next();
	});

	app.use(
		pinoHttp({
			logger,
			quietReqLogger: true,
			genReqId: (req) => req.id,
		})
	);

	app.use(json());

	app.use(
		OpenApiValidatorMiddleware({
			apiSpec,
			validateApiSpec: true,
			validateRequests: true,
			validateResponses: true,
		})
	);

	app['get']('/', (req, res) => {
		res.status(200).end();
	});

	app.use('/api/v1', await createVPICRoutes(dependencies.controllers.vpicController));

	app.use((req, _res, _next) => {
		throw new NotFound({ path: req.originalUrl });
	});

	app.use(errorHandlerMiddleware());

	return app;
};
