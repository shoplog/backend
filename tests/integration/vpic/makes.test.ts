import { Express } from 'express';
import { createApp } from 'src/api/setup';
import supertest from 'supertest';

describe('/vpic/makes', () => {
	const resourceUrl = '/api/v1/vpic/makes';
	let app: Express;

	beforeAll(async () => {
		app = await createApp();
	});

	describe('GET /:makeId/year/:year/models', () => {
		it('should respond with 200 Ok - all models for make and year', async () => {
			// Arrange
			const makeId = 448; // Toyota
			const year = 2002;

			// Act
			const body = await supertest(app)
				.get(`${resourceUrl}/${makeId}/year/${year}/models`)
				.expect(200)
				.then((res) => res.body);

			// Assert
			expect(body).toMatchSnapshot();
		});
	});
});
