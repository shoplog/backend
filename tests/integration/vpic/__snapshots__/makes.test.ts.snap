// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/vpic/makes GET /:makeId/year/:year/models should respond with 200 Ok - all models for make and year 1`] = `
[
  {
    "id": 2216,
    "name": "4-Runner",
  },
  {
    "id": 2465,
    "name": "Avalon",
  },
  {
    "id": 2469,
    "name": "Camry",
  },
  {
    "id": 2803,
    "name": "Camry Solara",
  },
  {
    "id": 3645,
    "name": "Celica",
  },
  {
    "id": 2208,
    "name": "Corolla",
  },
  {
    "id": 3650,
    "name": "Echo",
  },
  {
    "id": 2234,
    "name": "Forester",
  },
  {
    "id": 2213,
    "name": "Highlander",
  },
  {
    "id": 2235,
    "name": "Impreza",
  },
  {
    "id": 2211,
    "name": "Land Cruiser",
  },
  {
    "id": 3648,
    "name": "MR2",
  },
  {
    "id": 3649,
    "name": "Paseo",
  },
  {
    "id": 2209,
    "name": "Prius",
  },
  {
    "id": 2217,
    "name": "RAV4",
  },
  {
    "id": 2827,
    "name": "Sequoia",
  },
  {
    "id": 2468,
    "name": "Sienna",
  },
  {
    "id": 2223,
    "name": "Tacoma",
  },
  {
    "id": 2467,
    "name": "Tundra",
  },
]
`;
